jetty-server/src/main/java/org/eclipse/jetty/server/session/FileSessionDataStore.java

```diff
@@ -365,70 +376,70 @@ public void initializeStore()
-                            String existing = _sessionFileMap.putIfAbsent(sessionIdWithContext, filename);
-                            if (existing != null)
+                            //the session is for our context, populate the map with it
+                            String sessionIdWithContext = getIdWithContextFromFilename(filename);
+                            if (sessionIdWithContext != null)
                             {
-                                //if there was a prior filename, work out which has the most
-                                //recent modify time
-                                try
+                                //handle multiple session files existing for the same session: remove all
+                                //but the file with the most recent expiry time
+                                String existing = _sessionFileMap.putIfAbsent(sessionIdWithContext, filename);
+                                if (existing != null)
                                 {
-                                    long existingExpiry = getExpiryFromFilename(existing);
-                                    long thisExpiry = getExpiryFromFilename(filename);
-
-                                    if (thisExpiry > existingExpiry)
+                                    //if there was a prior filename, work out which has the most
+                                    //recent modify time
+                                    try
                                     {
-                                        //replace with more recent file
-                                        Path existingPath = _storeDir.toPath().resolve(existing);
-                                        //update the file we're keeping
-                                        _sessionFileMap.put(sessionIdWithContext, filename);
-                                        //delete the old file
-                                        Files.delete(existingPath);
-                                        if (LOG.isDebugEnabled())
-                                            LOG.debug("Replaced {} with {}", existing, filename);
+                                        long existingExpiry = getExpiryFromFilename(existing);
+                                        long thisExpiry = getExpiryFromFilename(filename);
+
+                                        if (thisExpiry > existingExpiry)
+                                        {
+                                            //replace with more recent file
+                                            Path existingPath = _storeDir.toPath().resolve(existing);
+                                            //update the file we're keeping
+                                            _sessionFileMap.put(sessionIdWithContext, filename);
+                                            //delete the old file
+                                            Files.delete(existingPath);
```

Comment:
This delete is still in here: if the intention is to avoid doing deletes inside the stream, you'll need to completely rework this code.

Reply:
oops - no worries, I have switched back to single walk anyway.